widget/tabcontainer.go

```diff
@@ -349,34 +296,97 @@ func (r *tabContainerRenderer) Refresh() {
 	r.line.FillColor = theme.ButtonColor()
 	r.line.Refresh()
 
-	for i, child := range r.container.Items {
-		tab := r.tabBar.Objects[i].(*tabButton)
-		tab.setText(child.Text)
+	if r.updateTabs() {
+		r.Layout(r.container.Size())
+	} else {
+		for i, button := range r.tabBar.Objects {
+			if i == r.container.current {
+				button.(*tabButton).Style = PrimaryButton
+			} else {
+				button.(*tabButton).Style = DefaultButton
+			}
 
-		old := r.objects[i]
-		if old == child.Content {
-			continue
+			button.Refresh()
 		}
+	}
+	canvas.Refresh(r.container)
+}
 
-		old.Hide()
-		r.objects[i] = child.Content
-		if i == r.container.current {
-			child.Content.Show()
-		} else {
-			child.Content.Hide()
+func (r *tabContainerRenderer) buildButton(item *TabItem, iconPos buttonIconPosition) *tabButton {
+	return &tabButton{
+		Text:         item.Text,
+		Icon:         item.Icon,
+		IconPosition: iconPos,
+		OnTap:        func() { r.container.SelectTab(item) },
+	}
+}
+
+func (r *tabContainerRenderer) buildTabBar(buttons []fyne.CanvasObject) *fyne.Container {
+	var lay fyne.Layout
+	if fyne.CurrentDevice().IsMobile() {
+		cells := len(buttons)
+		if cells == 0 {
+			cells = 1
+		}
+		lay = layout.NewGridLayout(cells)
+	} else if r.container.tabLocation == TabLocationLeading || r.container.tabLocation == TabLocationTrailing {
+		lay = layout.NewVBoxLayout()
+	} else {
+		lay = layout.NewHBoxLayout()
+	}
+
+	tabBar := fyne.NewContainerWithLayout(lay)
+	for _, button := range buttons {
+		tabBar.AddObject(button)
+	}
+	return tabBar
+}
+
+func (r *tabContainerRenderer) tabsInSync() bool {
+	if r.tabBar == nil {
+		return false
+	}
+	if r.tabLoc != r.container.tabLocation {
+		return false
+	}
+	if len(r.objects) != len(r.container.Items) {
+		return false
+	}
+	for i, item := range r.container.Items {
+		if item.Content != r.objects[i] {
+			return false
 		}
 	}
+	return true
+}
+
+func (r *tabContainerRenderer) updateTabs() bool {
+	if r.tabsInSync() {
+		return false
+	}
 
-	for i, button := range r.tabBar.Objects {
+	r.tabLoc = r.container.tabLocation
+	var iconPos buttonIconPosition
+	if fyne.CurrentDevice().IsMobile() || r.container.tabLocation == TabLocationLeading || r.container.tabLocation == TabLocationTrailing {
+		iconPos = buttonIconTop
+	} else {
+		iconPos = buttonIconInline
+	}
+	var buttons, objects []fyne.CanvasObject
+	for i, item := range r.container.Items {
+		button := r.buildButton(item, iconPos)
```

Comment:
Is it not worth re-using the existing buttons, if there are enough?

Reply:
I don’t think so.
1. A tab container usually doesn’t have more than a couple of tabs.
2. Modifying the tab list probably doesn’t happen that often.
3. Checking if an item is reusable introduces a complexity with limited use (see 1. and 2.)